core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1116,6 +1128,7 @@ class Log(@volatile var dir: File,
    */
   private def deleteOldSegments(predicate: (LogSegment, Option[LogSegment]) => Boolean, reason: String): Int = {
     lock synchronized {
+      checkIfLogOffline()
```

Comment:
It seems that we always need to check this after we acquire the lock. We could introduce a `withLock` method that did that so that we don't forget.

Reply:
@ijuma Strictly speaking we need to do `checkIfLogOffline` this iff we need to access disk content of this log. I am sure that we always need to grab the lock before accessing disk content of this log. But it is possible that the lock is grabbed but we don't try to access disk content. For example, `Log.activeProducersWithLastSequence()` grab lock but its operation is purely in-memory. It seems a bit weird to throw KafkaStorageException in this case, since KafkaStorageException indicates something goes wrong when broker accesses disk.

Reply:
That seems error prone. Once a Log is offline, we should make all of the operations in `Log` unavailable IMO. The fact that some may work just because they are relying on in memory state is a bit of an implementation detail and we should not rely on that. Is there a good reason to try and support some operations on `Log` after it's offline?

Reply:
@ijuma I agree that we should disable all log operation. I am just not sure we should throw KafkaStorageException due to the definition of KafkaStorageException. And the caller may not expect such exception as of now, thus more change and time is needed to make sure it works correctly if we want to disable all log operation.

Reply:
OK, can we file a JIRA for that then?

Reply:
@ijuma Certainly. I will file a JIRA for that. I am working on test now.

Reply:
I have created the JIRA KAFKA-6045 for this issue.

Reply:
Thanks.